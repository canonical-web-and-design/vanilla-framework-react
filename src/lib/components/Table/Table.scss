@import 'vanilla-framework/scss/base';
@import 'vanilla-framework/scss/patterns_table-sortable';
@import 'vanilla-framework/scss/utilities_content-align';
@include vf-p-table-sortable;
