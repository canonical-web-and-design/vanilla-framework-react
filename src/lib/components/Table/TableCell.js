import React from 'react';
import PropTypes from 'prop-types';

const TableCell = (props) => {
  const {
    align, ariaSort, children, columnHeader, onClick, role, rowHeader,
  } = props;
  const className = align ? `u-align--${align}` : null;

  if (columnHeader) {
    return (
      <th
        scope="col"
        role="columnheader"
        aria-sort={ariaSort}
        className={className}
        onClick={onClick}
      >
        {children}
      </th>);
  }

  if (rowHeader) {
    return (
      <th
        scope="row"
        role="rowheader"
        className={className}
      >
        {children}
      </th>);
  }

  return (
    <td
      role={role}
      className={className}
    >
      {children}
    </td>
  );
};

TableCell.defaultProps = {
  align: null,
  ariaSort: 'none',
  children: null,
  columnHeader: false,
  onClick: null,
  role: 'gridcell',
  rowHeader: false,
};

TableCell.propTypes = {
  align: PropTypes.string,
  ariaSort: PropTypes.string,
  children: PropTypes.node,
  columnHeader: PropTypes.bool,
  onClick: PropTypes.func,
  role: PropTypes.string,
  rowHeader: PropTypes.bool,
};

TableCell.displayName = 'TableCell';

export default TableCell;
