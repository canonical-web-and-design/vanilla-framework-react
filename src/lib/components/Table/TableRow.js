import React from 'react';
import PropTypes from 'prop-types';

const TableRow = (props) => {
  const tableRow = (
    <tr role="row">
      {props.children}
    </tr>
  );

  return tableRow;
};

TableRow.propTypes = {
  children: PropTypes.node.isRequired,
};

TableRow.displayName = 'TableRow';

export default TableRow;
