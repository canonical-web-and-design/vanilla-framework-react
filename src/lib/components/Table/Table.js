import React from 'react';
import PropTypes from 'prop-types';
import naturalCompare from 'natural-compare-lite';

import TableCell from './TableCell';
import TableRow from './TableRow';

class Table extends React.Component {
  constructor(props) {
    super(props);

    this.generateData = this.generateData.bind(this);
    this.generateHeader = this.generateHeader.bind(this);
    this.generateRows = this.generateRows.bind(this);
    this.getNewSortCondition = this.getNewSortCondition.bind(this);
    this.sortData = this.sortData.bind(this);

    this.state = {
      columns: props.columns ? [...props.columns] : null,
      data: props.data ? [...props.data] : null,
      originalData: props.data ? [...props.data] : null,
      sortCondition: props.sortCondition,
    };
  }

  componentWillMount() {
    const { columns, data, sortCondition } = this.state;

    // If Table is static, convert data to object
    if (!data || !columns) {
      this.generateData();
    }

    // Pre-sort Table if sortCondition prop is non-zero
    if (Object.keys(sortCondition).length === 1) {
      const columnKey = Object.keys(sortCondition)[0];
      this.sortData(columnKey);
    }
  }

  getNewSortCondition(columnKey) {
    const { sortCondition } = this.state;
    const newSortCondition = {};
    const options = ['ascending', 'descending', 'none'];

    // Move to next option in options array
    newSortCondition[columnKey] = (
      options[(options.indexOf(sortCondition[columnKey]) + 1) % options.length]
    );

    this.setState({ sortCondition: newSortCondition }, () => {
      this.sortData(columnKey);
    });
  }

  sortData(columnKey) {
    const { data, sortCondition } = this.state;

    if (sortCondition[columnKey] === 'none') {
      const unsortedData = this.state.originalData;
      this.setState({ data: [...unsortedData] });
    } else {
      const sortedData = data.sort((a, b) => {
        const x = a[columnKey];
        const y = b[columnKey];
        return (
          sortCondition[columnKey] === 'ascending' ? naturalCompare(x, y) : naturalCompare(y, x)
        );
      });

      this.setState({ data: sortedData });
    }
  }

  generateData() {
    const { children } = this.props;
    const columns = [];
    const data = [];

    children.forEach((row, rowIndex) => {
      if (row.props.tableHeader) {
        row.props.children.forEach((cell, cellIndex) => {
          let value = '';
          if (typeof cell.props.children === 'string') value = cell.props.children;
          columns.push({
            key: `column${cellIndex}`,
            label: value,
            align: cell.props.align,
            header: cell.props.columnHeader,
          });
        });
      } else {
        const index = rowIndex - 1;
        data[index] = {};
        data[index].id = rowIndex;
        row.props.children.forEach((cell, cellIndex) => {
          let value = '';
          if (typeof cell.props.children === 'string') value = cell.props.children;
          data[index][`column${cellIndex}`] = value;
        });
      }
    });

    this.setState({
      columns,
      data: [...data],
      originalData: data,
    });
  }

  generateHeader() {
    const { columns, sortCondition } = this.state;
    const { sortable } = this.props;

    const columnHeaders = columns.map((headerData) => {
      const ariaSort = sortCondition[headerData.key];
      const onClick = sortable ? () => this.getNewSortCondition(headerData.key) : null;
      return (
        <TableCell
          columnHeader
          key={headerData.key}
          ariaSort={ariaSort}
          align={headerData.align}
          onClick={onClick}
        >
          {headerData.label}
        </TableCell>);
    });

    const header = (
      <thead>
        <TableRow key="header-row">
          {columnHeaders}
        </TableRow>
      </thead>
    );

    return header;
  }

  generateRows() {
    const { data } = this.state;
    const rows = data.map((row) => {
      const cells = this.generateCells(row);
      return (
        <TableRow key={`row-${row.id}`}>
          {cells}
        </TableRow>
      );
    });

    return rows;
  }

  generateCells(row) {
    const { columns } = this.state;
    const cells = columns.map(headerData => (
      <TableCell
        rowHeader={headerData.header}
        key={`${headerData.key}-row-${row.id}`}
        align={headerData.align}
      >
        {row[headerData.key]}
      </TableCell>),
    );

    return cells;
  }

  render() {
    const { hasHeader, sortable } = this.props;
    const className = sortable ? 'p-table--sortable' : null;

    return (
      <table role="grid" className={className}>
        {hasHeader && this.generateHeader()}
        <tbody>
          {this.generateRows()}
        </tbody>
      </table>
    );
  }
}

Table.defaultProps = {
  children: null,
  columns: null,
  data: null,
  hasHeader: true,
  sortable: false,
  sortCondition: {},
};

Table.propTypes = {
  children: PropTypes.node,
  columns: PropTypes.arrayOf(PropTypes.object),
  data: PropTypes.arrayOf(PropTypes.object),
  hasHeader: PropTypes.bool,
  sortable: PropTypes.bool,
  sortCondition: PropTypes.object,
};

Table.displayName = 'Table';

export default Table;
