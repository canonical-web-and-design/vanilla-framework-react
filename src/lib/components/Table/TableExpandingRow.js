import React from 'react';
import PropTypes from 'prop-types';

import TableRow from './TableRow';
import TableCell from './TableCell';

class TableExpandingRow extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      showPanel: false,
    };

    this.togglePanel = this.togglePanel.bind(this);
  }

  togglePanel() {
    this.setState({
      showPanel: !this.state.showPanel,
    });
  }

  render() {
    const { showPanel } = this.state;
    const { data, children } = this.props;
    const tableRow = (
      <TableRow>
        {data.cellData.map(datum => (
          <TableCell key={`tabel-cell-${datum.content}`} align={datum.align}>{datum.content}</TableCell>
        ))}

        <TableCell align="right">
          <button
            className="u-toggle"
            aria-controls={`toggle-panel-${data.id}`}
            aria-expanded={this.state.showPanel}
            onClick={this.togglePanel}
          >
            {showPanel ? 'Hide' : 'Show'}
          </button>
        </TableCell>
        <td id={`toggle-panel-${data.id}`} className="p-table-expanding__panel" aria-hidden={!this.state.showPanel}>
          {children}
        </td>
      </TableRow>
    );

    return tableRow;
  }
}

TableExpandingRow.propTypes = {
  children: PropTypes.node.isRequired,
  data: PropTypes.object.isRequired,
};

TableExpandingRow.displayName = 'TableExpandingRow';

export default TableExpandingRow;
