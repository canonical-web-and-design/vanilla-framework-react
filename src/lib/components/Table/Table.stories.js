import React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean } from '@storybook/addon-knobs';
import { withInfo } from '@storybook/addon-info';

import Table from './Table';
import TableCell from './TableCell';
import TableRow from './TableRow';

const columns = [
  [
    { key: 'header', label: '', header: true },
    { key: 'column1', label: 'Header' },
    { key: 'column2', label: 'Header' },
    { key: 'column3', label: 'Header', align: 'right' },
  ],
  [
    { key: 'name', label: 'Name' },
    { key: 'users', label: 'Users', align: 'right' },
    { key: 'units', label: 'Units', align: 'right' },
    { key: 'revenue', label: 'Revenue', align: 'right' },
  ],
];

const data = [
  [{
    id: 0, header: 'Header row', column1: 'Table cell', column2: 'Table cell', column3: 111,
  }, {
    id: 1, header: 'Header row', column1: 'Table cell', column2: 'Table cell', column3: 222,
  }, {
    id: 2, header: 'Header row', column1: 'Table cell', column2: 'Table cell', column3: 333,
  }],
  [{
    id: 0, name: 'Grape', users: 8, units: 19, revenue: '$70',
  }, {
    id: 1, name: 'Melon', users: 12, units: 23, revenue: '$99',
  }, {
    id: 2, name: 'Apple', users: 9, units: 17, revenue: '$120',
  }],
];

storiesOf('Table', module)
  .add('Static Table',
    withInfo('The Table component can be constructed in two ways - either manually using appropriate sub-components (TableRow and TableCell), or dynamically using two object props (data and columns). This is an example of a Static Table built manually.')(() => (
      <Table
        hasHeader={boolean('Header', true)}
        sortable={boolean('Sortable', false)}
      >
        <TableRow tableHeader>
          <TableCell columnHeader />
          <TableCell>Header</TableCell>
          <TableCell>Header</TableCell>
          <TableCell align="right">Header</TableCell>
        </TableRow>
        <TableRow>
          <TableCell>Header row</TableCell>
          <TableCell>Table cell</TableCell>
          <TableCell>Table cell</TableCell>
          <TableCell>111</TableCell>
        </TableRow>
        <TableRow>
          <TableCell>Header row</TableCell>
          <TableCell>Table cell</TableCell>
          <TableCell>Table cell</TableCell>
          <TableCell>222</TableCell>
        </TableRow>
        <TableRow>
          <TableCell>Header row</TableCell>
          <TableCell>Table cell</TableCell>
          <TableCell>Table cell</TableCell>
          <TableCell>333</TableCell>
        </TableRow>
      </Table>),
    ),
  )

  .add('Object Table',
    withInfo('An Object Table is built using object props for the data as well as for the column headers.')(() => (
      <Table
        hasHeader={boolean('Header', true)}
        sortable={boolean('Sortable', false)}
        columns={columns[0]}
        data={data[0]}
      />),
    ),
  )

  .add('No Header',
    withInfo('Tables can have the header row disabled.')(() => (
      <Table
        hasHeader={boolean('Header', false)}
        sortable={boolean('Sortable', false)}
        columns={columns[0]}
        data={data[0]}
      />),
    ),
  )

  .add('Sortable',
    withInfo('Adding a "sortable" prop to a Table allows each column to cycle through ascending, descending and no sort.')(() => (
      <Table
        hasHeader={boolean('Header', true)}
        sortable={boolean('Sortable', true)}
        columns={columns[1]}
        data={data[1]}
      />),
    ),
  )

  .add('Pre-sorted',
    withInfo('A Table can be pre-sorted if a valid sortCondition prop is provided. For Object Tables the sortCondition object key is the column key. In Static Tables, it is column(index) e.g. { column0: ascending }')(() => (
      <Table
        hasHeader={boolean('Header', true)}
        sortable={boolean('Sortable', true)}
        sortCondition={{ units: 'descending' }}
        columns={columns[1]}
        data={data[1]}
      />),
    ),
  );
