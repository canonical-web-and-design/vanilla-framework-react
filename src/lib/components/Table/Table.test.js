import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import Table from './Table';
import TableCell from './TableCell';
import TableRow from './TableRow';
import TableExpandingRow from './TableExpandingRow';

const columns = [
  [
    { key: 'header', label: '', header: true },
    { key: 'column1', label: 'Header' },
    { key: 'column2', label: 'Header' },
    { key: 'column3', label: 'Header', align: 'right' },
  ],
  [
    { key: 'name', label: 'Name' },
    { key: 'users', label: 'Users', align: 'right' },
    { key: 'units', label: 'Units', align: 'right' },
    { key: 'revenue', label: 'Revenue', align: 'right' },
  ],
  [
    { key: 'name', label: 'Name' },
    { key: 'users', label: 'Users', align: 'right' },
    { key: 'units', label: 'Units', align: 'right' },
    { key: 'actions', label: 'Actions', align: 'right' },
  ],
];

const data = [
  [{
    id: 0, header: 'Header row', column1: 'Table cell', column2: 'Table cell', column3: 111,
  },
  {
    id: 1, header: 'Header row', column1: 'Table cell', column2: 'Table cell', column3: 222,
  },
  {
    id: 2, header: 'Header row', column1: 'Table cell', column2: 'Table cell', column3: 333,
  }],
  [{
    id: 0, name: 'Grape', users: 8, units: 19, revenue: '$70',
  },
  {
    id: 1, name: 'Melon', users: 12, units: 23, revenue: '$99',
  },
  {
    id: 2, name: 'Apple', users: 9, units: 17, revenue: '$120',
  }],
  [{
    id: 0,
    cellData: [{
      content: 'Grape',
      align: 'left',
    }, {
      content: 8,
      align: 'right',
    }, {
      content: 18,
      align: 'right',
    }, {
      content: '$70',
      align: 'right',
    }],
  }, {
    id: 1,
    cellData: [{
      content: 'Melon',
      align: 'left',
    }, {
      content: 12,
      align: 'right',
    }, {
      content: 23,
      align: 'right',
    }, {
      content: '$99',
      align: 'right',
    }],
  }, {
    id: 2,
    cellData: [{
      content: 'Apple',
      align: 'left',
    }, {
      content: 9,
      align: 'right',
    }, {
      content: 17,
      align: 'right',
    }, {
      content: '$120',
      align: 'right',
    }],
  }],
];

describe('Table component', () => {
  it('should render a static table correctly', () => {
    const table = ReactTestRenderer.create(
      <Table>
        <TableRow tableHeader>
          <TableCell columnHeader />
          <TableCell>Header</TableCell>
          <TableCell>Header</TableCell>
          <TableCell align="right">Header</TableCell>
        </TableRow>
        <TableRow>
          <TableCell>Header row</TableCell>
          <TableCell>Table cell</TableCell>
          <TableCell>Table cell</TableCell>
          <TableCell>111</TableCell>
        </TableRow>
        <TableRow>
          <TableCell>Header row</TableCell>
          <TableCell>Table cell</TableCell>
          <TableCell>Table cell</TableCell>
          <TableCell>222</TableCell>
        </TableRow>
        <TableRow>
          <TableCell>Header row</TableCell>
          <TableCell>Table cell</TableCell>
          <TableCell>Table cell</TableCell>
          <TableCell>333</TableCell>
        </TableRow>
      </Table>,
    );
    const json = table.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render a dynamic table correctly', () => {
    const table = ReactTestRenderer.create(
      <Table
        columns={columns[0]}
        data={data[0]}
      />,
    );
    const json = table.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render a headerless table correctly', () => {
    const table = ReactTestRenderer.create(
      <Table
        hasHeader={false}
        columns={columns[0]}
        data={data[0]}
      />,
    );
    const json = table.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render a sortable table correctly', () => {
    const table = ReactTestRenderer.create(
      <Table
        sortable
        columns={columns[1]}
        data={data[1]}
      />,
    );
    const json = table.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render a pre-sorted table correctly', () => {
    const table = ReactTestRenderer.create(
      <Table
        sortCondition={{ units: 'descending' }}
        columns={columns[1]}
        data={data[1]}
      />,
    );
    const json = table.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render an expanding table correctly', () => {
    const tableExpanding = ReactTestRenderer.create(
      <Table
        columns={columns[2]}
        hasHeader
        expanding
      >
        {data[2].map(datum => (
          <TableExpandingRow key={`expanding-row-${datum.id}`} data={datum}>
            <div>Hello world!</div>
          </TableExpandingRow>
        ))}
      </Table>,
    );
    const json = tableExpanding.toJSON();
    expect(json).toMatchSnapshot();
  });
});
