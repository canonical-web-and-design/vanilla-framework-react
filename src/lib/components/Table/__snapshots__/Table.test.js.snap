// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component should render a dynamic table correctly 1`] = `
<table
  className={null}
  role="grid"
>
  <thead>
    <tr
      role="row"
    >
      <th
        aria-sort="none"
        className={null}
        onClick={null}
        role="columnheader"
        scope="col"
      >
        
      </th>
      <th
        aria-sort="none"
        className={null}
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Header
      </th>
      <th
        aria-sort="none"
        className={null}
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Header
      </th>
      <th
        aria-sort="none"
        className="u-align--right"
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Header
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      role="row"
    >
      <th
        className={null}
        role="rowheader"
        scope="row"
      >
        Header row
      </th>
      <td
        className={null}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className={null}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        111
      </td>
    </tr>
    <tr
      role="row"
    >
      <th
        className={null}
        role="rowheader"
        scope="row"
      >
        Header row
      </th>
      <td
        className={null}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className={null}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        222
      </td>
    </tr>
    <tr
      role="row"
    >
      <th
        className={null}
        role="rowheader"
        scope="row"
      >
        Header row
      </th>
      <td
        className={null}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className={null}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        333
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table component should render a headerless table correctly 1`] = `
<table
  className={null}
  role="grid"
>
  <tbody>
    <tr
      role="row"
    >
      <th
        className={null}
        role="rowheader"
        scope="row"
      >
        Header row
      </th>
      <td
        className={null}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className={null}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        111
      </td>
    </tr>
    <tr
      role="row"
    >
      <th
        className={null}
        role="rowheader"
        scope="row"
      >
        Header row
      </th>
      <td
        className={null}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className={null}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        222
      </td>
    </tr>
    <tr
      role="row"
    >
      <th
        className={null}
        role="rowheader"
        scope="row"
      >
        Header row
      </th>
      <td
        className={null}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className={null}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        333
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table component should render a pre-sorted table correctly 1`] = `
<table
  className={null}
  role="grid"
>
  <thead>
    <tr
      role="row"
    >
      <th
        aria-sort="none"
        className={null}
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Name
      </th>
      <th
        aria-sort="none"
        className="u-align--right"
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Users
      </th>
      <th
        aria-sort="descending"
        className="u-align--right"
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Units
      </th>
      <th
        aria-sort="none"
        className="u-align--right"
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Revenue
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      role="row"
    >
      <td
        className={null}
        role="gridcell"
      >
        Melon
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        12
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        23
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        $99
      </td>
    </tr>
    <tr
      role="row"
    >
      <td
        className={null}
        role="gridcell"
      >
        Grape
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        8
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        19
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        $70
      </td>
    </tr>
    <tr
      role="row"
    >
      <td
        className={null}
        role="gridcell"
      >
        Apple
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        9
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        17
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        $120
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table component should render a sortable table correctly 1`] = `
<table
  className="p-table--sortable"
  role="grid"
>
  <thead>
    <tr
      role="row"
    >
      <th
        aria-sort="none"
        className={null}
        onClick={[Function]}
        role="columnheader"
        scope="col"
      >
        Name
      </th>
      <th
        aria-sort="none"
        className="u-align--right"
        onClick={[Function]}
        role="columnheader"
        scope="col"
      >
        Users
      </th>
      <th
        aria-sort="none"
        className="u-align--right"
        onClick={[Function]}
        role="columnheader"
        scope="col"
      >
        Units
      </th>
      <th
        aria-sort="none"
        className="u-align--right"
        onClick={[Function]}
        role="columnheader"
        scope="col"
      >
        Revenue
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      role="row"
    >
      <td
        className={null}
        role="gridcell"
      >
        Grape
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        8
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        19
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        $70
      </td>
    </tr>
    <tr
      role="row"
    >
      <td
        className={null}
        role="gridcell"
      >
        Melon
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        12
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        23
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        $99
      </td>
    </tr>
    <tr
      role="row"
    >
      <td
        className={null}
        role="gridcell"
      >
        Apple
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        9
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        17
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        $120
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table component should render a static table correctly 1`] = `
<table
  className={null}
  role="grid"
>
  <thead>
    <tr
      role="row"
    >
      <th
        aria-sort="none"
        className={null}
        onClick={null}
        role="columnheader"
        scope="col"
      >
        
      </th>
      <th
        aria-sort="none"
        className={null}
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Header
      </th>
      <th
        aria-sort="none"
        className={null}
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Header
      </th>
      <th
        aria-sort="none"
        className="u-align--right"
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Header
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      role="row"
    >
      <th
        className={null}
        role="rowheader"
        scope="row"
      >
        Header row
      </th>
      <td
        className={null}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className={null}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        111
      </td>
    </tr>
    <tr
      role="row"
    >
      <th
        className={null}
        role="rowheader"
        scope="row"
      >
        Header row
      </th>
      <td
        className={null}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className={null}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        222
      </td>
    </tr>
    <tr
      role="row"
    >
      <th
        className={null}
        role="rowheader"
        scope="row"
      >
        Header row
      </th>
      <td
        className={null}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className={null}
        role="gridcell"
      >
        Table cell
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        333
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table component should render an expanding table correctly 1`] = `
<table
  className="p-table-expanding"
  role="grid"
>
  <thead>
    <tr
      role="row"
    >
      <th
        aria-sort="none"
        className={null}
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Name
      </th>
      <th
        aria-sort="none"
        className="u-align--right"
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Users
      </th>
      <th
        aria-sort="none"
        className="u-align--right"
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Units
      </th>
      <th
        aria-sort="none"
        className="u-align--right"
        onClick={null}
        role="columnheader"
        scope="col"
      >
        Actions
      </th>
      <td
        className={null}
        role="gridcell"
      />
    </tr>
  </thead>
  <tbody>
    <tr
      role="row"
    >
      <td
        className="u-align--left"
        role="gridcell"
      >
        Grape
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        8
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        18
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        $70
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        <button
          aria-controls="toggle-panel-0"
          aria-expanded={false}
          className="u-toggle"
          onClick={[Function]}
        >
          Show
        </button>
      </td>
      <td
        aria-hidden={true}
        className="p-table-expanding__panel"
        id="toggle-panel-0"
      >
        <div>
          Hello world!
        </div>
      </td>
    </tr>
    <tr
      role="row"
    >
      <td
        className="u-align--left"
        role="gridcell"
      >
        Melon
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        12
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        23
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        $99
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        <button
          aria-controls="toggle-panel-1"
          aria-expanded={false}
          className="u-toggle"
          onClick={[Function]}
        >
          Show
        </button>
      </td>
      <td
        aria-hidden={true}
        className="p-table-expanding__panel"
        id="toggle-panel-1"
      >
        <div>
          Hello world!
        </div>
      </td>
    </tr>
    <tr
      role="row"
    >
      <td
        className="u-align--left"
        role="gridcell"
      >
        Apple
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        9
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        17
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        $120
      </td>
      <td
        className="u-align--right"
        role="gridcell"
      >
        <button
          aria-controls="toggle-panel-2"
          aria-expanded={false}
          className="u-toggle"
          onClick={[Function]}
        >
          Show
        </button>
      </td>
      <td
        aria-hidden={true}
        className="p-table-expanding__panel"
        id="toggle-panel-2"
      >
        <div>
          Hello world!
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;
